### PR TITLE
test: schema support for testprovider

### PR DESCRIPTION
**Pull request by EronWright**
# Description

This PR enhances the testprovider to have a Pulumi schema, e.g. to be able to use it with Pulumi YAML.

Related to https://github.com/pulumi/pulumi/issues/15788

## Checklist

- [x] I have run `make tidy` to update any new dependencies
- [x] I have run `make lint` to verify my code passes the lint check
  - [ ] I have formatted my code using `gofumpt`

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

File: tests/testprovider/echo.go [hunk 1]
```diff
@@ -20,12 +20,40 @@ import (
...
+		IsComponent: false,
+		ObjectTypeSpec: pschema.ObjectTypeSpec{
+			IsOverlay:   false,
+			Description: "A test resource that echoes its input.",
+			Properties: map[string]pschema.PropertySpec{
+				"echo": {
+					TypeSpec: pschema.TypeSpec{
+						Ref: "pulumi.json#/Resource",
+					},
+					Description: "A resource to echo.",
```
Comment (justinvp) [inline tests/testprovider/echo.go:41]: Shouldn't this be `Ref: "pulumi.json#/Any"` and the same for the input, to align with existing usage?

https://github.com/pulumi/pulumi/blob/4a012def4b2191e3b98f5fa5e36c1eb3c4c3334e/tests/integration/construct_component/testcomponent-python/echo.py#L10

File: tests/testprovider/echo.go [hunk 2]
```diff
@@ -20,12 +20,40 @@ import (
...
+	pschema "github.com/pulumi/pulumi/pkg/v3/codegen/schema"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/plugin"
 	rpc "github.com/pulumi/pulumi/sdk/v3/proto/go"
 
 	"google.golang.org/protobuf/types/known/emptypb"
 )
 
+func init() {
+	providerSchema.Resources["testprovider:index:Echo"] = pschema.ResourceSpec{
+		IsComponent: false,
```
Comment (justinvp) [inline tests/testprovider/echo.go:32]: Is this line necessary?

File: tests/testprovider/echo.go [hunk 3]
```diff
@@ -20,12 +20,40 @@ import (
...
+		ObjectTypeSpec: pschema.ObjectTypeSpec{
+			IsOverlay:   false,
```
Comment (justinvp) [inline tests/testprovider/echo.go:34]: Is this line necessary?

File: tests/testprovider/echo.go (hunk 1 above)
Comment (EronWright) [reply tests/testprovider/echo.go:41]: You're right, I didn't realize that the Echo resource was able to echo non-resources.

**Review by AaronFriel — APPROVED** [review]

**Review by Frassle — APPROVED** [review]

**Pull request merged**